source/mir/combinatorics/package.d

```diff
@@ -359,6 +366,10 @@ struct Permutations
         this.state = state;
 
         _empty = indices.length == 0;
+
+        // factorial
+        foreach (i; 1..indices.length + 1)
+            _max_states *= i;
```

Comment:
all the other ranges provide a `length` function - i don't think it makes a noticeable overhead. Alternatively we could also compute it on-demand only